mongodb/docs-worker-pool
# DOP-3340: Initial implementation of OAS page builder module

### Ticket

[DOP-3340](https://jira.mongodb.org/browse/DOP-3340)

### Testing

Follow the steps below if you're interested in testing locally. It's also advised to read the README added in this PR for additional context and/or instructions.

1) Use the parser (with the [following commit](https://github.com/mongodb/snooty-parser/commit/144386d522c8693dc4a02424ef7786ed0bd133ff)) to parse a docs repo that has one or more OpenAPI content pages. Examples include `cloud-docs` and `docs-app-services`. (Using the aforementioned commit instead of the `HEAD` is important here due to changes and experimentation of ideal metadata outputs. The module will not work if used with the latest commits.)
2) Clone the team's [fork of Redoc](https://github.com/mongodb-forks/redoc). Set up by installing dependencies for both the root folder (Redoc) and within the `cli/` folder (Redoc CLI).
3) Build and compile Redoc and Redoc CLI by running `npm run bundle` in the root directory of the Redoc fork. IF there are any issues building, then you might need to use `npm link ../` to link the local copy of `redoc` as the CLI's dependency. `npm run bundle` should include a build for the CLI, which will be output as `cli/index.js`.
4) (Optional) Use the frontend to build the docs repo you parsed. All files should be built to the `public/` directory. The autobuilder module should output the generated HTML files from Redoc to the same `public/` directory output from Snooty. However, for testing purposes, you could always just have it output to wherever you want.
5) Install the dependencies of this autobuilder module and run using the following command:

```
npm run build && npm run start -- --bundle <path-to-zipped-output-from-parser> --output <path-to-target-directory> --redoc <path-to-local-js-file-for-redoc-cli> --repo <path-to-docs-repo-directory-you-parsed>
```

### Notes

* Adds initial implementation for the new OpenAPI page builder module.
* Not yet callable or usable by the autobuilder; this functionality should be added in [DOP-3341](https://jira.mongodb.org/browse/DOP-3341). The module is usable locally for testing purposes (see: testing instructions above). If having difficulties with testing locally and/or if the instructions are unclear/inaccurate, please feel free to ping me!
